src/magic.cpp

```diff
@@ -0,0 +1,804 @@
+                                       type->duration_increment ) ),
+                     type->max_duration );
+}
+
+void spell::gain_level()
+{
+    gain_exp( exp_to_next_level() );
+}
+/*
+void spell::lose_level()
+{
+    if( level > 0 ) {
+        level--;
+    }
+}
+*/
+bool spell::is_max_level() const
+{
+    return get_level() >= type->max_level;
+}
+
+bool spell::can_learn() const
+{
+    if( type->spell_class == trait_id( "NONE" ) ) {
+        return true;
+    }
+    return owner->has_trait( type->spell_class );
+}
+
+int spell::energy_cost() const
+{
+    // todo: formula to vary energy cost
+    return type->base_energy_cost;
+}
+
+bool spell::can_cast() const
+{
+    if( !owner->knows_spell( type->id ) ) {
+        // how in the world can this happen?
+        debugmsg( "ERROR: owner of spell does not know spell" );
+        return false;
+    }
+    switch( type->energy_source ) {
+        case mana_energy:
+            return owner->available_mana() >= energy_cost();
+        case stamina_energy:
+            return owner->stamina >= energy_cost();
+        case hp_energy: {
+            for( int i = 0; i < num_hp_parts; i++ ) {
+                if( energy_cost() < owner->hp_cur[i] ) {
+                    return true;
+                }
+            }
+        }
+        case bionic_energy:
+            return owner->power_level >= energy_cost();
+        case none_energy:
+        default:
+            return true;
+    }
+}
+
+int spell::modded_int() const
+{
+    return owner->int_cur;
+}
+
+int spell::spellcraft_level() const
+{
+    return owner->get_skill_level( skill_id( "spellcraft" ) );
+}
+
+int spell::get_difficulty() const
+{
+    return type->difficulty;
+}
+
+int spell::casting_time() const
+{
+    // todo: formula for casting time
+    return type->base_casting_time;
+}
+
+std::string spell::name() const
+{
+    return _( type->name );
+}
+
+float spell::spell_fail() const
+{
+    // formula is based on the following:
+    // exponential curve
+    // effective skill of 0 or less is 100% failure
+    // effective skill of 8 (8 int, 0 spellcraft, 0 spell level, spell difficulty 0) is ~50% failure
+    // effective skill of 30 is 0% failure
+    const float effective_skill = 2 * ( get_level() - get_difficulty() ) + modded_int() +
+                                  spellcraft_level();
+    // add an if statement in here because sufficiently large numbers will definitely overflow because of exponents
+    if( effective_skill > 30.0f ) {
+        return 0.0f;
+    } else if( effective_skill < 0.0f ) {
+        return 1.0f;
+    }
+    const float fail_chance = pow( ( effective_skill - 30.0f ) / 30.0f, 2 );
+    return clamp( fail_chance, 0.0f, 1.0f );
+}
+
+std::string spell::colorized_fail_percent() const
+{
+    const float fail_fl = spell_fail() * 100.0f;
+    std::string fail_str;
+    fail_fl == 100.0f ? fail_str = _( "Difficult!" ) : fail_str = string_format( "%.1f %s %s", fail_fl,
```

Comment:
This formatting is strange.  If you want a literal `%` you can get that with `%%`; you don't need another `%s` argument.  Also, I'd suggest putting `Failure Chance` in the format string, and translating that, rather than just translating the one component.  That way another language can put the percentage at the end if it makes more sense for them.

Reply:
```cpp
std::string fail_str = fail_fl == 100.0f ? _( "Difficult!" ) : string_format( "%.1f%% failure chance.", fail_fl );
```

Reply:
i kept getting errors when i tried that, but i'll change it back.